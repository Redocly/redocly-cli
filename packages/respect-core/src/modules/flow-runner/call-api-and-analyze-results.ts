import { checkCriteria } from './success-criteria/index.js';
import { checkSchema } from './schema/index.js';
import { CHECKS } from '../checks/index.js';
import { createRuntimeExpressionCtx } from './context/index.js';
import { evaluateRuntimeExpressionPayload } from '../runtime-expressions/index.js';

import type { RequestData } from './prepare-request.js';
import type { TestContext, Step } from '../../types.js';

// TODO: split into two functions
export async function callAPIAndAnalyzeResults({
  ctx,
  workflowId,
  step,
  requestData,
}: {
  ctx: TestContext;
  workflowId: string;
  step: Step;
  requestData: RequestData;
}) {
  // clear checks in case of retry
  step.checks = [];

  const checksResult = {
    successCriteriaCheck: true,
    schemaCheck: true,
    networkCheck: true,
  };

  try {
    step.response = await ctx.apiClient.fetchResult({ ctx, step, requestData, workflowId });
  } catch (error: any) {
    step.checks.push({
      name: CHECKS.NETWORK_ERROR,
      passed: false,
      message: error.message,
      severity: ctx.severity['NETWORK_ERROR'],
    });
    checksResult.networkCheck = false;
    return checksResult;
  }

  const request = ctx.$workflows[workflowId].steps[step.stepId].request;

  step.verboseLog = ctx.apiClient.getVerboseResponseLogs();

  if (step.successCriteria) {
    const successCriteriaChecks = checkCriteria({
      workflowId,
      step,
      criteria: step.successCriteria,
      ctx: {
        ...ctx,
        $request: request,
        $response: step.response,
        $inputs: ctx.$workflows[workflowId].inputs,
      },
    });

    checksResult.successCriteriaCheck = successCriteriaChecks.every(
      (check) => check.passed || ['off', 'warn'].includes(check.severity)
    );
    step.checks.push(...successCriteriaChecks);
  }

  const schemaChecks = checkSchema({
    stepCallCtx: {
      $request: request,
      $response: step.response,
      $inputs: ctx.$workflows[workflowId].inputs,
    },
    descriptionOperation: requestData.openapiOperation,
    ctx,
  });

  if (schemaChecks.length) {
    checksResult.schemaCheck = schemaChecks.every(
      (check) => check.passed || ['off', 'warn'].includes(check.severity)
    );
    step.checks.push(...schemaChecks);
  }

  // store step level outputs
  const outputs: Record<string, any> = {};
  if (step.outputs) {
    const runtimeExpressionContext = createRuntimeExpressionCtx({
      ctx: {
        ...ctx,
        $request: request,
        $response: step.response,
        $inputs: ctx.$workflows[workflowId].inputs,
      },
      workflowId,
      step,
    });

    for (const outputKey of Object.keys(step.outputs)) {
      outputs[outputKey] = evaluateRuntimeExpressionPayload({
        payload: step.outputs[outputKey],
        context: runtimeExpressionContext,
      });
    }
  }

  // save local $steps context
  ctx.$steps[step.stepId] = {
    outputs: { ...ctx.$steps[step.stepId].outputs, ...outputs },
  };
  // save $workflows context
  ctx.$workflows[workflowId].steps[step.stepId] = {
    outputs: { ...ctx.$steps[step.stepId].outputs, ...outputs },
    request,
    response: step.response,
  };

  return checksResult;
}
