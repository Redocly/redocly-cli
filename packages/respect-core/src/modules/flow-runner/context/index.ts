export * from './create-test-context';
export * from './set-public-steps';
export * from './set-public-workflows';
export * from './create-runtime-expression-ctx';
