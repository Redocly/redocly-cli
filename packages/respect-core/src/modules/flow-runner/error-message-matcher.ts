import { bold } from 'colorette';

export function errorMessageMatcher(
  hint: string, // assertion returned from call to matcherHint
  generic: string, // condition which correct value must fulfill
  specific?: string // incorrect value returned from call to printWithType
): string {
  return `${hint}\n\n${bold('Matcher error')}: ${generic}${
    typeof specific === 'string' ? `\n\n${specific}` : ''
  }`;
}
