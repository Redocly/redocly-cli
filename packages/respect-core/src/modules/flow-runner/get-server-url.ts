import { getValueFromContext } from '../context-parser/index.js';
import { formatCliInputs } from './inputs/index.js';

import type { ExtendedOperation, TestContext } from '../../types.js';
import type { OperationDetails } from '../description-parser/index.js';

export type GetServerUrlInput = {
  ctx: TestContext;
  descriptionName?: string;
  openapiOperation?: (OperationDetails & Record<string, any>) | undefined;
  xOperation?: ExtendedOperation;
};

export type ServerObject = {
  url: string;
  description?: string;
  variables?: Record<
    string,
    {
      default: string;
      enum?: string[];
      description?: string;
    }
  >;
};

export function getServerUrl({
  ctx,
  descriptionName,
  openapiOperation,
  xOperation,
}: GetServerUrlInput): { url: string } | undefined {
  if (!descriptionName && xOperation?.url) {
    return { url: xOperation?.url };
  }

  // Handle server overrides from command line `server` option
  if (ctx.options?.server && descriptionName) {
    const serverOverrides = formatCliInputs(ctx.options.server);
    if (serverOverrides[descriptionName]) {
      return { url: serverOverrides[descriptionName] };
    }
  }

  const sourceDescription = ctx.sourceDescriptions?.find((sd) => sd.name === descriptionName);

  if (
    sourceDescription &&
    sourceDescription.type === 'openapi' &&
    sourceDescription['x-serverUrl']
  ) {
    if (sourceDescription['x-serverUrl'].startsWith('http')) {
      return {
        url: sourceDescription['x-serverUrl'],
      };
    }

    const serverIndexInDescription = sourceDescription['x-serverUrl'].startsWith('$servers.')
      ? sourceDescription['x-serverUrl'].split('.')[1]
      : undefined;

    if (!serverIndexInDescription) {
      return undefined;
    }

    const serverObject = getValueFromContext(
      `$sourceDescriptions.${descriptionName}.servers.${serverIndexInDescription}`,
      ctx
    );

    return resolveOpenApiServerUrlWithVariables(serverObject);
  }

  if (openapiOperation?.servers?.[0]) {
    const activeSourceDescription = ctx.sourceDescriptions?.find(
      (sd) => sd.name === openapiOperation.sourceDescriptionName
    );
    let serverUrlOverride;

    if (ctx.sourceDescriptions?.length === 1 && ctx.sourceDescriptions[0]) {
      serverUrlOverride =
        'x-serverUrl' in ctx.sourceDescriptions[0]
          ? ctx.sourceDescriptions[0]['x-serverUrl']
          : undefined;
    } else if (activeSourceDescription) {
      serverUrlOverride =
        'x-serverUrl' in activeSourceDescription
          ? activeSourceDescription['x-serverUrl']
          : undefined;
    }

    return serverUrlOverride
      ? { url: serverUrlOverride }
      : resolveOpenApiServerUrlWithVariables(openapiOperation.servers[0]);
  }

  if (!descriptionName && ctx?.sourceDescriptions && ctx.sourceDescriptions.length === 1) {
    const sourceDescription = ctx.sourceDescriptions[0];

    if ('x-serverUrl' in sourceDescription && sourceDescription['x-serverUrl']) {
      return { url: sourceDescription['x-serverUrl'] };
    } else {
      return (
        resolveOpenApiServerUrlWithVariables(
          ctx.$sourceDescriptions[sourceDescription.name]?.servers[0]
        ) || undefined
      );
    }
  }

  if (
    !descriptionName ||
    !ctx.$sourceDescriptions ||
    !ctx.$sourceDescriptions[descriptionName] ||
    !ctx.$sourceDescriptions[descriptionName]?.servers?.length
  ) {
    return undefined;
  }

  // Get first available server url from the description
  return resolveOpenApiServerUrlWithVariables(
    ctx.$sourceDescriptions[descriptionName].servers?.[0]
  );
}

function resolveOpenApiServerUrlWithVariables(server?: ServerObject) {
  if (!server) {
    return undefined;
  }
  return {
    url: server.url,
    parameters: Object.entries(server.variables || {})
      .map(([key, value]) => ({
        in: 'path',
        name: key,
        value: value.default,
      }))
      .filter(({ value }) => !!value),
  };
}
