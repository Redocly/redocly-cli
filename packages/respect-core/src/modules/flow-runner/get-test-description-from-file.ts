import { bold, red } from 'colorette';
import { getTotals, formatProblems, lint, bundle, createConfig } from '@redocly/openapi-core';
import { type CollectFn } from '@redocly/openapi-core';
import * as path from 'node:path';
import { existsSync } from 'node:fs';
import { printConfigLintTotals } from '../../utils/cli-outputs.js';
import { isTestFile } from '../../utils/file.js';
import { readYaml } from '../../utils/yaml.js';
import { version } from '../../utils/package.js';

export async function bundleArazzo(filePath: string, collectSpecData?: CollectFn) {
  const fileName = path.basename(filePath);

  if (!fileName) {
    throw new Error('Invalid file name');
  }

  if (!existsSync(filePath)) {
    const relativePath = path.relative(process.cwd(), filePath);
    throw new Error(
      `Could not find source description file '${fileName}' at path '${relativePath}'`
    );
  }

  const fileContent = await readYaml(filePath);

  if (!isTestFile(fileName, fileContent)) {
    throw new Error(
      `No test files found. File ${fileName} does not follows naming pattern "*.[yaml | yml | json]" or have not valid "Arazzo" description.`
    );
  }

  const config = await createConfig({
    extends: ['recommended-strict'],
    arazzo1Rules: {
      'no-criteria-xpath': 'error',
      'respect-supported-versions': 'warn',
      'no-x-security-scheme-name-without-openapi': 'error',
      'x-security-scheme-required-values': 'error',
      'no-x-security-scheme-name-in-workflow': 'error',
    },
  });

  const lintProblems = await lint({
    ref: filePath,
    config,
  });

  if (lintProblems.length) {
    const fileTotals = getTotals(lintProblems);

    formatProblems(lintProblems, {
      totals: fileTotals,
      version,
    });

    printConfigLintTotals(fileTotals);
  }

  const bundledDocument = await bundle({
    ref: filePath,
    config,
    dereference: true,
  });

  collectSpecData?.(bundledDocument.bundle.parsed || {});

  const errorLintProblems = lintProblems.filter((problem) => problem.severity === 'error');
  if (errorLintProblems.length) {
    throw new Error(`${red('Found errors in Arazzo description')} ${bold(fileName)}`);
  }

  return bundledDocument.bundle.parsed || {};
}
