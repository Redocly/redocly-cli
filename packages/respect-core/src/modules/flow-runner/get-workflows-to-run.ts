import { red, yellow } from 'colorette';
import { DefaultLogger } from '../../utils/logger/logger.js';

import type { Workflow } from '../../types.js';

const logger = DefaultLogger.getInstance();

export function getWorkflowsToRun(
  workflows: Workflow[],
  workflowsToRun: string[] | undefined,
  workflowsToSkip: string[] | undefined
): Workflow[] {
  let filteredWorkflows: Workflow[] = [];

  if (workflowsToRun && workflowsToRun.length) {
    filteredWorkflows = filterWorkflowsToRun(workflows, workflowsToRun);
  } else if (workflowsToSkip && workflowsToSkip.length) {
    filteredWorkflows = filterWorkflowsToSkip(workflows, workflowsToSkip);
  } else {
    filteredWorkflows = workflows;
  }

  return filteredWorkflows;
}

function filterWorkflowsToSkip(workflows: Workflow[], workflowsToSkip: string[]) {
  const workflowsToRun = workflows.filter(
    (workflow) => !workflowsToSkip.includes(workflow.workflowId)
  );

  if (!workflowsToRun.length) {
    logger.log(`${red('All workflows are skipped')}`);
    logger.printNewLine();
    return [];
  }

  logger.log(`${yellow(`Following workflows are skipped: ${workflowsToSkip.join(', ')}`)}`);
  logger.printNewLine();

  return workflowsToRun;
}

function filterWorkflowsToRun(workflows: Workflow[], workflowsToRun: string[]) {
  const filteredWorkflows = filterWorkflowsByIds(workflows, workflowsToRun);

  if (!filteredWorkflows.length) {
    throw new Error(`Following workflows don't exist: ${workflowsToRun.join(', ')}`);
  }

  if (filteredWorkflows.length === workflowsToRun.length) {
    return filteredWorkflows;
  } else {
    const notExistingWorkflows = workflowsToRun.filter((workflowId) => {
      return !workflows.find((workflow) => workflow.workflowId === workflowId);
    });

    logger.log(`Following workflows don't exist: ${notExistingWorkflows.join(', ')}`);
    return filteredWorkflows;
  }
}

function filterWorkflowsByIds(workflows: Workflow[], workflowIds: string[]) {
  return workflows.filter((workflow) => workflowIds.includes(workflow.workflowId));
}
