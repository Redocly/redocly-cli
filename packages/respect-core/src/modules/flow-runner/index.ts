export * from './runner.js';
export * from './context/index.js';
export * from './get-workflows-to-run.js';
export * from './schema/index.js';
export * from './inputs/index.js';
export * from './outputs/index.js';
export * from '../checks/index.js';
export * from './error-message-matcher.js';
export * from './success-criteria/index.js';
export * from './get-server-url.js';
export * from './call-api-and-analyze-results.js';
export * from './run-step.js';
export * from './prepare-request.js';
export * from './read-env-variables.js';
export * from './resolve-running-workflows.js';
