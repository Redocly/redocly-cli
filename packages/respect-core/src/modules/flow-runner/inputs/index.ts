export * from './format-cli-inputs.js';
export * from './map-input-values-to-schema.js';
