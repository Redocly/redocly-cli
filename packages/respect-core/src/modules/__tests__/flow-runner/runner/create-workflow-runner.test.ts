import type { ApiFetcher } from '../../../../utils/api-fetcher';
import type { Workflow, TestContext } from '../../../../types';

import { runWorkflow, DEFAULT_SEVERITY_CONFIGURATION } from '../../../flow-runner';
import { DefaultLogger } from '../../../../utils/logger/logger';
import { type ResponseContext } from '../../../../types';

const logger = DefaultLogger.getInstance();
describe('runWorkflow', () => {
  const fileName = 'test.yaml';
  it('should run workflow', async () => {
    const apiClient = {
      setDescriptionParameters: vi.fn(),
      getRequestHeaderParams: vi.fn(),
      getRequestParams: vi.fn(),
      setDefaultsParameters: vi.fn(),
      setDisconnectDescription: vi.fn(),
      setPath: vi.fn(),
      setDescriptionRequestBody: vi.fn(),
      setDescriptionResponses: vi.fn(),
      setDescriptionContentType: vi.fn(),
      setParameters: vi.fn(),
      setTestCaseRequestBody: vi.fn(),
      setMethod: vi.fn(),
      setApiBase: vi.fn(),
      fetchResult: vi.fn(),
      getVerboseResponseLogs: vi.fn(),
    } as unknown as ApiFetcher;

    vi.mocked(apiClient.fetchResult).mockResolvedValue({
      statusCode: 200,
    } as ResponseContext);

    const workflow = {
      workflowId: 'test',
      steps: [
        {
          stepId: 'test',
          'x-operation': {
            url: 'http://localhost:3000/test',
            method: 'GET',
          },
          successCriteria: [
            {
              condition: '$statusCode == 200',
            },
          ],
        },
      ],
    } as unknown as Workflow;

    const ctx = {
      executedSteps: [],
      apiClient,
      workflows: [workflow],
      $workflows: {
        test: {
          inputs: {},
          steps: {
            test: {
              successCriteria: [
                {
                  condition: '$statusCode == 200',
                },
              ],
              response: {
                statusCode: 200,
              },
            },
          },
        },
      },
      $steps: {
        test: {
          successCriteria: [
            {
              condition: '$statusCode == 200',
            },
          ],
          response: {
            statusCode: 200,
          },
        },
      },
      severity: DEFAULT_SEVERITY_CONFIGURATION,
      sourceDescriptions: [
        {
          name: 'test',
          type: 'openapi',
          url: 'openapi.yaml',
        },
      ],
      options: {
        verbose: false,
        workflowPath: fileName,
      },
    } as unknown as TestContext;

    await runWorkflow({ workflowInput: 'test', ctx });

    expect(apiClient.fetchResult).toBeCalled();
  });

  it('should return if no steps', async () => {
    const apiClient = {
      fetchResult: vi.fn(),
    } as unknown as ApiFetcher;

    const workflow = {
      workflowId: 'test',
      steps: [],
    } as unknown as Workflow;

    const ctx = {
      apiClient,
      workflows: [workflow],
      options: {
        verbose: false,
        workflowPath: fileName,
      },
    } as unknown as TestContext;

    await runWorkflow({ workflowInput: 'test', ctx });

    expect(apiClient.fetchResult).not.toBeCalled();
  });

  it('should throw an error if no workflow exists', async () => {
    const apiClient = {
      fetchResult: vi.fn(),
    } as unknown as ApiFetcher;

    const ctx = {
      apiClient,
      workflows: [],
      options: {
        verbose: false,
        workflowPath: fileName,
      },
    } as unknown as TestContext;
    await expect(runWorkflow({ workflowInput: 'test', ctx })).rejects.toThrowError();
    expect(apiClient.fetchResult).not.toBeCalled();
  });

  it('should set workflow outputs', async () => {
    const apiClient = {
      fetchResult: vi.fn(),
      getVerboseResponseLogs: vi.fn(),
    } as unknown as ApiFetcher;

    vi.mocked(apiClient.fetchResult).mockResolvedValue({
      statusCode: 200,
    } as ResponseContext);

    const workflow = {
      workflowId: 'test',
      outputs: {
        test: 'test',
      },

      steps: [
        {
          stepId: 'test',
          'x-operation': {
            url: 'http://localhost:3000/test',
            method: 'GET',
          },
          successCriteria: [
            {
              condition: '$statusCode == 200',
            },
          ],
        },
      ],
    } as unknown as Workflow;

    const ctx = {
      apiClient,
      workflows: [workflow],
      executedSteps: [],
      $workflows: {
        test: {
          outputs: {},
          inputs: {},
          steps: {
            test: {
              successCriteria: [
                {
                  condition: '$statusCode == 200',
                },
              ],
              response: {
                statusCode: 200,
              },
            },
          },
        },
      },
      $steps: {
        test: {
          successCriteria: [
            {
              condition: '$statusCode == 200',
            },
          ],
          response: {
            statusCode: 200,
          },
        },
      },
      sourceDescriptions: [
        {
          name: 'test',
          type: 'openapi',
          url: 'openapi.yaml',
        },
      ],
      severity: DEFAULT_SEVERITY_CONFIGURATION,
      options: {
        verbose: false,
        workflowPath: fileName,
      },
      $outputs: {},
    } as unknown as TestContext;

    await runWorkflow({ workflowInput: 'test', ctx });

    expect(ctx.$outputs?.test).toEqual({ test: 'test' });
    expect(ctx.$workflows.test.outputs).toEqual({ test: 'test' });
  });

  it('should return if workflow does not have steps', async () => {
    const apiClient = {
      fetchResult: vi.fn(),
    } as unknown as ApiFetcher;

    vi.mocked(apiClient.fetchResult).mockResolvedValue({
      statusCode: 200,
    } as ResponseContext);

    const workflow = {
      workflowId: 'test',
      outputs: {
        test: 'test',
      },

      steps: [],
    } as unknown as Workflow;

    const ctx = {
      apiClient,
      workflows: [workflow],
      $workflows: {
        test: {
          outputs: {},
          inputs: {},
          steps: {
            test: {
              successCriteria: [
                {
                  condition: '$statusCode == 200',
                },
              ],
              response: {
                statusCode: 200,
              },
            },
          },
        },
      },
      $steps: {
        test: {
          successCriteria: [
            {
              condition: '$statusCode == 200',
            },
          ],
          response: {
            statusCode: 200,
          },
        },
      },
      sourceDescriptions: [
        {
          name: 'test',
          type: 'openapi',
          url: 'openapi.yaml',
        },
      ],
      options: {
        verbose: false,
        workflowPath: fileName,
      },
      $outputs: {},
    } as unknown as TestContext;

    await runWorkflow({ workflowInput: 'test', ctx });

    expect(apiClient.fetchResult).not.toBeCalled();
  });

  it('should run workflow within step execution', async () => {
    const mockLogger = vi.spyOn(logger, 'log').mockImplementation(() => {});
    const apiClient = {
      fetchResult: vi.fn(),
    } as unknown as ApiFetcher;

    vi.mocked(apiClient.fetchResult).mockResolvedValue({
      statusCode: 200,
    } as ResponseContext);

    const workflow = {
      workflowId: 'test',
      outputs: {
        test: 'test',
      },
      steps: [],
    } as unknown as Workflow;

    const ctx = {
      apiClient,
      workflows: [workflow],
      $workflows: {
        test: {
          outputs: {},
          inputs: {},
          steps: {},
        },
        parentWorkflowId: {
          outputs: {},
          inputs: {},
          steps: {},
        },
      },
      $steps: {},
      sourceDescriptions: [
        {
          name: 'test',
          type: 'openapi',
          url: 'openapi.yaml',
        },
      ],
      options: {
        verbose: false,
        workflowPath: fileName,
      },
      $outputs: {},
    } as unknown as TestContext;

    await runWorkflow({
      workflowInput: 'test',
      ctx,
    });

    expect(mockLogger).toMatchSnapshot();

    mockLogger.mockRestore();
  });

  it('should accept workflow as an input', async () => {
    const apiClient = {
      fetchResult: vi.fn(),
      getVerboseResponseLogs: vi.fn(),
    } as unknown as ApiFetcher;

    vi.mocked(apiClient.fetchResult).mockResolvedValue({
      statusCode: 200,
    } as ResponseContext);

    const workflow = {
      workflowId: 'test',
      outputs: {
        test: 'test',
      },

      steps: [
        {
          stepId: 'test',
          'x-operation': {
            url: 'http://localhost:3000/test',
            method: 'GET',
          },
          successCriteria: [
            {
              condition: '$statusCode == 200',
            },
          ],
        },
      ],
    } as unknown as Workflow;

    const ctx = {
      apiClient,
      workflows: [workflow],
      executedSteps: [],
      $workflows: {
        test: {
          outputs: {
            test: 'test',
          },

          inputs: {},
          steps: {
            test: {
              successCriteria: [
                {
                  condition: '$statusCode == 200',
                },
              ],
              response: {
                statusCode: 200,
              },
            },
          },
        },
        parentWorkflowId: {
          inputs: {},
          outputs: {},
          steps: {},
        },
      },
      $steps: {
        test: {
          successCriteria: [
            {
              condition: '$statusCode == 200',
            },
          ],
          response: {
            statusCode: 200,
          },
        },
      },
      sourceDescriptions: [
        {
          name: 'test',
          type: 'openapi',
          url: 'openapi.yaml',
        },
      ],
      severity: DEFAULT_SEVERITY_CONFIGURATION,
      options: {
        verbose: false,
        workflowPath: fileName,
      },
      $outputs: {},
    } as unknown as TestContext;

    await runWorkflow({
      workflowInput: workflow,
      ctx,
    });

    expect(ctx.$outputs?.test).toEqual({ test: 'test' });
    expect(ctx.$workflows.test.outputs).toEqual({ test: 'test' });
  });
});
