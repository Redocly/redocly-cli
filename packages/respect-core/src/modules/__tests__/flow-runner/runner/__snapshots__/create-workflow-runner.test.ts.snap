// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runWorkflow > should run workflow within step execution 1`] = `
[MockFunction log] {
  "calls": [
    [
      "  [1mRunning workflow[22m [34mtest.yaml / test[39m",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
